### PR TITLE
Store migration for 3.0 enterprise format

### PR DESCRIPTION
**Pull request by tinwelint**
Although it's more than that. In this PR the store migration is made a bit more generic,
for two reasons:
- To make future migrations easier to implement
- Community (where the migration lives) doesn't know about all formats anymore

The legacy stores have been removed in favor of simply using `NeoStores`
injected with an older format to read the legacy store.

### CODE REVIEW

File: community/neo4j/src/test/java/upgrade/StoreUpgradeOnStartupTest.java [hunk 1]
```diff
@@ -78,10 +78,10 @@
     public static Collection<Object[]> versions()
     {
         return Arrays.asList(
-                new Object[]{Legacy20Store.LEGACY_VERSION},
-                new Object[]{Legacy21Store.LEGACY_VERSION},
-                new Object[]{Legacy22Store.LEGACY_VERSION},
-                new Object[]{Legacy23Store.LEGACY_VERSION}
+                new Object[]{LowLimitV2_0.STORE_VERSION},
```
Comment (MishaDemianenko) [inline community/neo4j/src/test/java/upgrade/StoreUpgradeOnStartupTest.java]: Each parameter can be plain `String`, there is no need to wrap it into `Object[]`

File: community/neo4j/src/test/java/upgrade/StoreUpgraderInterruptionTestIT.java [hunk 2]
```diff
@@ -83,10 +84,10 @@
     public static Collection<Object[]> versions()
     {
         return Arrays.asList(
-                new Object[]{Legacy20Store.LEGACY_VERSION},
-                new Object[]{Legacy21Store.LEGACY_VERSION},
-                new Object[]{Legacy22Store.LEGACY_VERSION},
-                new Object[]{Legacy23Store.LEGACY_VERSION}
+                new Object[]{LowLimitV2_0.STORE_VERSION},
```
Comment (MishaDemianenko) [inline community/neo4j/src/test/java/upgrade/StoreUpgraderInterruptionTestIT.java]: Each parameter can be plain String, there is no need to wrap it into Object[]

File: community/neo4j/src/test/java/upgrade/StoreUpgraderTest.java [hunk 3]
```diff
@@ -126,10 +128,10 @@ public StoreUpgraderTest( String version )
     public static Collection<Object[]> versions()
     {
         return Arrays.asList(
-                new Object[]{Legacy20Store.LEGACY_VERSION},
-                new Object[]{Legacy21Store.LEGACY_VERSION},
-                new Object[]{Legacy22Store.LEGACY_VERSION},
-                new Object[]{Legacy23Store.LEGACY_VERSION}
+                new Object[]{LowLimitV2_0.STORE_VERSION},
```
Comment (MishaDemianenko) [inline community/neo4j/src/test/java/upgrade/StoreUpgraderTest.java]: Each parameter can be plain String, there is no need to wrap it into Object[]

File: community/kernel/src/main/java/org/neo4j/graphdb/factory/GraphDatabaseSettings.java [hunk 4]
```diff
@@ -355,36 +356,40 @@ private static String defaultPageCacheMemory()
...
-            "bytes. " +
-            "This means that if the block size is 120, the size of the stored records will be 128 bytes.")
+            "Note that each character in a string occupies two bytes, meaning that e.g a block size of 120 will hold " +
+            "a 60 character long string before overflowing into a second block. " +
+            "Also note that each block carries a ~10B of overhead so record size on disk will be slightly larger " +
+            "than the configured block size" )
     @Internal
-    public static final Setting<Integer> string_block_size = setting("string_block_size", INTEGER, "120",min(1));
+    public static final Setting<Integer> string_block_size = setting("string_block_size", INTEGER,
+            dynamicRecordDataSizeForAligningWith( 128 ), min(1) );
```
Comment (MishaDemianenko) [inline community/kernel/src/main/java/org/neo4j/graphdb/factory/GraphDatabaseSettings.java]: looks like in all updated properties using 1 as min is not reasonable

File: community/kernel/src/main/java/org/neo4j/helpers/ArrayUtil.java [hunk 5]
```diff
@@ -485,6 +486,27 @@ public static boolean isEmpty( Object[] array )
...
     }
 
+    /**
+     * Filters an array, only including the items that passes the {@code filter}.
+     *
+     * @param array items to filter.
+     * @param filter {@link Predicate} to match each item.
+     * @return a new array with the items that passed the filter.
+     */
+    public static <T> T[] filter( T[] array, Predicate<T> filter )
```
Comment (MishaDemianenko) [inline community/kernel/src/main/java/org/neo4j/helpers/ArrayUtil.java]: In both cases where this method is used same can be achieved using `Stream.of` or `Arrays.stream` 
using plain stream api.
Thats why i think we do not need this method

File: community/kernel/src/main/java/org/neo4j/kernel/impl/storemigration/participant/StoreMigrator.java [hunk 6]
```diff
@@ -177,6 +182,15 @@ public void migrate( File storeDir, File migrationDir, MigrationProgressMonitor.
         // must be a proper translation happening while doing so.
     }
 
+    private void migrateWithDirectMigration( File storeDir, File migrationDir,
+            RecordFormats oldFormat, RecordFormats newFormat ) throws IOException
+    {
+        DirectRecordStoreMigrator migrator = new DirectRecordStoreMigrator( pageCache, fileSystem, config );
+        //Not interested in migrating MetaData store.
+        StoreType[] stores = ArrayUtil.filter( StoreType.values(), type -> type != StoreType.META_DATA );
```
Comment (MishaDemianenko) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/storemigration/participant/StoreMigrator.java]: This can be done using pure stream api:
`Arrays.stream(StoreType.values())
                .filter( storeType -> storeType != StoreType.META_DATA )
                .toArray( StoreType[]::new );`

File: community/kernel/src/main/java/org/neo4j/kernel/impl/store/format/BaseRecordFormats.java [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+    @Override
+    public boolean hasCapability( Capability capability )
+    {
+        return contains( capabilities(), capability );
+    }
+
+    @Override
+    public boolean hasSameCapabilities( RecordFormats other, int types )
+    {
+        Capability[] myFormatCapabilities =
```
Comment (MishaDemianenko) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/store/format/BaseRecordFormats.java]: can be achieved using plain stream api and `Arrays.equals` without usages of custom `filter` and `contains` methods

File: community/kernel/src/main/java/org/neo4j/kernel/impl/store/format/Capability.java [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+     * Lucene version 3.x
+     */
+    LUCENE_3( Capability.TYPE_INDEX ),
+
+    /**
+     * Lucene version 5.x
+     */
+    LUCENE_5( Capability.TYPE_INDEX );
+
+    public static final int TYPE_FORMAT = 0x1;
```
Comment (MishaDemianenko) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/store/format/Capability.java]: lets introduce separate CapabilityType enum, that will be much more readable and easy to provide correct parameter for isType method for example

File: community/kernel/src/main/java/org/neo4j/kernel/impl/store/format/Capability.java [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+    public static final int TYPE_STORE = 0x2;
+    public static final int TYPE_INDEX = 0x4;
+    public static final int ALL_TYPES = TYPE_FORMAT | TYPE_STORE | TYPE_INDEX;
```
Comment (MishaDemianenko) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/store/format/Capability.java]: not used

File: community/kernel/src/main/java/org/neo4j/kernel/impl/store/format/RecordFormats.java [hunk 10]
```diff
@@ -58,5 +68,22 @@ public Factory( String key, String... altKeys )
 
     RecordFormat<RelationshipTypeTokenRecord> relationshipTypeToken();
 
-    RecordFormat<RelationshipGroupRecord> relationshipGroup();
+    RecordFormat<DynamicRecord> dynamic();
+
+    boolean hasStore( StoreType store );
```
Comment (MishaDemianenko) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/store/format/RecordFormats.java]: do not think this method has something to do with RecordFormats, since we have no notions of stores here and should not operate them

File: community/kernel/src/main/java/org/neo4j/kernel/impl/store/CommonAbstractStore.java [hunk 11]
```diff
@@ -689,6 +696,57 @@ protected long scanForHighId()
...
+        try
+        {
+            for ( int i = 0; i < expectedVersionBytes.length; )
+            {
+                // If the version bytes are bigger than record size then we must also compare with subsets
+                // of those bytes in recordSize chunks
+                boolean mismatch = false;
+                for ( int j = 0; i < expectedVersionBytes.length && j < recordSize; i++, j++ )
+                {
+                    byte b = cursor.getByte( offset + j );
```
Comment (MishaDemianenko) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/store/CommonAbstractStore.java:731]: if expectedVersionBytes.length > recordSize inner for loop start to behave strangely: we will no try to read from second record because of `cursor.getByte( offset + j )` that will start again and `!mismatch`
but i do not know if that is a realistic case :)

File: community/kernel/src/main/java/org/neo4j/kernel/impl/storemigration/participant/StoreMigrator.java [hunk 12]
```diff
@@ -177,6 +182,15 @@ public void migrate( File storeDir, File migrationDir, MigrationProgressMonitor.
...
+        migrator.migrate( storeDir, oldFormat, migrationDir, newFormat, stores, new StoreType[0] );
```
Comment (MishaDemianenko) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/storemigration/participant/StoreMigrator.java]: last param can be represented ad vararg so you can avoid useless new StoreType[0]

File: community/kernel/src/main/java/org/neo4j/kernel/impl/store/format/RecordFormats.java [hunk 13]
```diff
@@ -44,9 +48,15 @@ public Factory( String key, String... altKeys )
 
     String storeVersion();
 
+    /**
+     * @return format generation, with the intent of usage being that a store can migrate to a newer or
+     * same generation, but not to an older generation.
+     */
+    int generation();
```
Comment (MishaDemianenko) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/store/format/RecordFormats.java:61]: i find it hard to follow and understand why we need it in addition to versions of particular formats, especially when generation following completely independent from numbering scheme, it's really confusing

File: community/kernel/src/test/java/org/neo4j/kernel/impl/storemigration/legacystore/v20/Legacy20Store.java [hunk 14]
```diff
@@ -19,31 +19,11 @@
...
-
-import static org.hamcrest.CoreMatchers.containsString;
-import static org.junit.Assert.assertThat;
-import static org.junit.Assert.fail;
-
-import static org.neo4j.kernel.impl.storemigration.legacystore.v20.Legacy20Store.assertLegacyAndCurrentVersionHaveSameLength;
-
-public class Legacy20StoreTest
+/**
+ * Here as a place holder for using {@link Unzip} to find old db tarballs.
```
Comment (MishaDemianenko) [inline community/kernel/src/test/java/org/neo4j/kernel/impl/storemigration/legacystore/v20/Legacy20Store.java:25]: that is really strange reason to keep an empty interface

File: community/kernel/src/main/java/org/neo4j/kernel/impl/storemigration/FileOperation.java [hunk 15]
```diff
@@ -44,13 +44,17 @@
         @Override
         public void perform( FileSystemAbstraction fs, String fileName,
                 File fromDirectory, boolean skipNonExistentFromFile,
-                File toDirectory, boolean overwrite )
+                File toDirectory, ExistingTargetStrategy existingTargetStrategy )
```
Comment (MishaDemianenko) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/storemigration/FileOperation.java:47]: i think it should be useful to have separate test for this utility class

File: community/kernel/src/test/java/org/neo4j/kernel/impl/storemigration/UpgradableDatabaseTest.java [hunk 16]
```diff
@@ -244,10 +259,9 @@ public void shouldCommunicateWhatCausesInabilityToUpgrade() throws IOException
             catch ( StoreUpgrader.UnexpectedUpgradingStoreVersionException e )
             {
                 // then
-                assertThat( e.getMessage(), containsString( neostoreFilename ) );
-                assertThat( e.getMessage(), containsString(
-                        "has a store version number that we cannot upgrade from. Expected '" +
-                        Legacy23Store.LEGACY_VERSION + "' but file is version '" + version + "'." ) );
+                File expectedFile = new File( workingDirectory, neostoreFilename ).getAbsoluteFile();
+                assertEquals( String.format( MESSAGE, expectedFile, version ), e.getMessage() );
+                assertThat( e.getMessage(), containsString( version ) );
```
Comment (MishaDemianenko) [inline community/kernel/src/test/java/org/neo4j/kernel/impl/storemigration/UpgradableDatabaseTest.java]: no reason to check that message contains version twice

**Comment by MishaDemianenko** [comment]
@tinwelint there are still commented out lines that should be uncommented as soon as migration ready and ignored tests for the same reason

**Comment by MishaDemianenko** [comment]
new version of migration does not look like easier to read and follow even some step are so much better

**Comment by MishaDemianenko** [comment]
@tinwelint my first round of reviews completed

File: community/kernel/src/main/java/org/neo4j/kernel/impl/store/format/BaseRecordFormats.java (hunk 7 above)
Comment (tinwelint) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/store/format/BaseRecordFormats.java]: We don't want to be dependent on items in the array being in the exact same order though.

File: community/kernel/src/main/java/org/neo4j/kernel/impl/storemigration/participant/StoreMigrator.java (hunk 6 above)
Comment (tinwelint) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/storemigration/participant/StoreMigrator.java]: How about we implement ArrayUtil#filter using the above... it's certainly much more convenient that using that big chunk of code every time you want to filter an array

File: community/kernel/src/main/java/org/neo4j/kernel/impl/store/format/RecordFormats.java (hunk 10 above)
Comment (tinwelint) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/store/format/RecordFormats.java]: Only the format can have that information, who else would know that?

RecordFormats and store version is conflated (as in the rest of the code historically as well). At some point we might break those two apart, but I don't think it's necessary to do that right now.

File: community/kernel/src/main/java/org/neo4j/kernel/impl/store/CommonAbstractStore.java (hunk 11 above)
Comment (tinwelint) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/store/CommonAbstractStore.java:731]: It is a realistic case, for example node store has a bigger trailer than record size, and it works for that store as well. Have you spotted an issue? Can it be exposed with a test?

File: community/kernel/src/main/java/org/neo4j/kernel/impl/storemigration/participant/StoreMigrator.java (hunk 12 above)
Comment (tinwelint) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/storemigration/participant/StoreMigrator.java]: yup

File: community/kernel/src/main/java/org/neo4j/kernel/impl/store/format/RecordFormats.java (hunk 13 above)
Comment (tinwelint) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/store/format/RecordFormats.java:61]: it's to be able to tell whether or not we're trying to do an upgrade or a downgrade... given two formats. Is there a better way?

File: community/kernel/src/test/java/org/neo4j/kernel/impl/storemigration/legacystore/v20/Legacy20Store.java (hunk 14 above)
Comment (tinwelint) [inline community/kernel/src/test/java/org/neo4j/kernel/impl/storemigration/legacystore/v20/Legacy20Store.java:25]: Yes, but let's change this in another PR

File: community/kernel/src/main/java/org/neo4j/kernel/impl/storemigration/FileOperation.java (hunk 15 above)
Comment (tinwelint) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/storemigration/FileOperation.java:47]: Yes, let's add that in another PR to not hold this one hanging due to that

**Comment by tinwelint** [comment]
@MishaDemianenko which commented lines and ignored tests have you spotted?

File: community/kernel/src/main/java/org/neo4j/kernel/impl/store/format/RecordFormats.java (hunk 10 above)
Comment (MishaDemianenko) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/store/format/RecordFormats.java]: This used only from one place in `DirectRecordStoreMigrator` and from that place it seems more logically to have it in `NeoStores` to know if that store was applicable based on capabilities of available format

File: community/kernel/src/main/java/org/neo4j/kernel/impl/store/CommonAbstractStore.java (hunk 11 above)
Comment (MishaDemianenko) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/store/CommonAbstractStore.java:731]: fair, will try to spot it

File: community/kernel/src/main/java/org/neo4j/kernel/impl/store/format/RecordFormats.java (hunk 13 above)
Comment (MishaDemianenko) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/store/format/RecordFormats.java:61]: since our formats has version as part of class name we can have same version inside of them, with exactly same value stored in a version field and do comparison based on those instead

**Comment by MishaDemianenko** [comment]
@tinwelint i knew you will gonna ask them, but i did not track/mark them in anyway, will try to use IDE power to find those

**Comment by MishaDemianenko** [comment]
@tinwelint found ignored test at StoreUpgradeIntegrationTest # migrationShouldFail

**Comment by tinwelint** [comment]
@MishaDemianenko good because that last commit removed that ignore

File: community/kernel/src/main/java/org/neo4j/kernel/impl/store/format/RecordFormats.java (hunk 10 above)
Comment (tinwelint) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/store/format/RecordFormats.java]: Oh, right that method came about before capabilities existed. I agree with this. Lemme make this change.

Comment (tinwelint) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/store/format/RecordFormats.java]: Oh actually, that thing isn't even needed anymore since the direct store migrator will not used between formats that aren't logically compatible, which implies that they have the same set of stores as well.

File: community/kernel/src/main/java/org/neo4j/kernel/impl/store/format/RecordFormats.java (hunk 13 above)
Comment (tinwelint) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/store/format/RecordFormats.java:61]: How does `HighLimit` enter into that? I think it would be error prone to try and infer it, better be explicit about it

Comment (MishaDemianenko) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/store/format/RecordFormats.java:61]: to include those guys we will need more complex schema with base version (neo version) and subversion for formats at lest - that sound complex, agree, so just please add how those numbers for generations need to be chosen for future formats into javadocs and lets leave it as it is
